java-checks/src/main/java/org/sonar/java/checks/security/SecureXmlTransformerCheck.java

```diff
@@ -0,0 +1,128 @@
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+
+import static org.sonar.java.checks.helpers.ConstantUtils.resolveAsStringConstant;
+import static org.sonar.java.matcher.TypeCriteria.subtypeOf;
+
+@Rule(key = "S4435")
+public class SecureXmlTransformerCheck extends AbstractMethodDetection {
+
+  private static final String TRANSFORMER_FACTORY_CLASS_NAME = TransformerFactory.class.getName();
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return Collections.singletonList(
+      MethodMatcher.create()
+        .typeDefinition(TRANSFORMER_FACTORY_CLASS_NAME)
+        .name("newInstance")
+        .withAnyParameters());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree methodInvocation) {
+    Tree enclosingMethod = enclosingMethod(methodInvocation);
+    if (enclosingMethod == null) {
+      return;
+    }
+    MethodBodyVisitor visitor = new MethodBodyVisitor();
+    enclosingMethod.accept(visitor);
+    if (!visitor.foundCallsToSecuringMethods()) {
+      reportIssue(methodInvocation.methodSelect(), "Secure this \"Transformer\" by either disabling external DTDs or enabling secure processing.");
+    }
+  }
+
+  @CheckForNull
+  private static Tree enclosingMethod(Tree tree) {
+    Tree parent = tree.parent();
+    while (!parent.is(Kind.CLASS, Kind.METHOD)) {
+      parent = parent.parent();
+    }
+    if (parent.is(Kind.CLASS)) {
+      return null;
+    }
+    return parent;
+  }
+
+  private static class MethodBodyVisitor extends BaseTreeVisitor {
+
+    private static final MethodMatcher SET_FEATURE =
+      MethodMatcher.create()
+        .typeDefinition(subtypeOf(TRANSFORMER_FACTORY_CLASS_NAME))
+        .name("setFeature")
+        .parameters("java.lang.String", "boolean");
+
+    private static final MethodMatcher SET_ATTRIBUTE =
+      MethodMatcher.create()
+        .typeDefinition(subtypeOf(TRANSFORMER_FACTORY_CLASS_NAME))
+        .name("setAttribute")
+        .parameters("java.lang.String", "java.lang.Object");
+
+    private boolean hasSecureProcessingFeature = false;
+    private boolean hasSecuredExternalDtd = false;
+    private boolean hasSecuredExternalStylesheet = false;
+
+    private boolean foundCallsToSecuringMethods() {
+      return hasSecureProcessingFeature || (hasSecuredExternalDtd && hasSecuredExternalStylesheet);
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree methodInvocation) {
+      Arguments arguments = methodInvocation.arguments();
+
+      if (SET_FEATURE.matches(methodInvocation)
+        && XMLConstants.FEATURE_SECURE_PROCESSING.equals(resolveAsStringConstant(arguments.get(0)))
+        && LiteralUtils.isTrue(arguments.get(1))) {
+
+        hasSecureProcessingFeature = true;
+      }
+
+      if (SET_ATTRIBUTE.matches(methodInvocation)) {
+        String attributeName = resolveAsStringConstant(arguments.get(0));
+        String attributeValue = resolveAsStringConstant(arguments.get(1));
```

Comment:
could also use `JavaPropertiesHelper.retrievedPropertyDefaultValue` ?

Reply:
I don't know which behavior we would want for this rule when properties are used and I don't think it would add value to the implementation.